ops/testing.py

```diff
@@ -859,6 +867,13 @@ def add_relation(self, relation_name: str, remote_app: str, *,
             if unit_data is not None:
                 self.update_relation_data(relation_id, remote_unit, unit_data)
 
+        # If we don't already have a network binding for this relation id, create one.
+        if not self._backend._networks.get((relation_name, relation_id)):
+            self.add_network("10.0.0.10", endpoint=relation_name, relation_id=relation_id)
+        # If we don't already have a default network binding for this endpoint, create one.
+        if not self._backend._networks.get((relation_name, None)):
+            self.add_network("192.0.2.0", endpoint=relation_name)
```

Comment:
Sorry, I put this in a comment but should have put it in a review as well so that it was clearer: I think that this needs to have a `if not self._backend._networks.get((None, None)):` guard around it, to handle the case where there's a default network that will be used when there is not one specific to the relation name+id or relation name. The comment has details of a couple of charms where not doing this would change the behaviour, and (from what I understand) not in a way that we expect.